marcdown_project/frontend/src/components/ModalPermissions.js

```diff
@@ -0,0 +1,144 @@
+import React, { Component } from "react";
+
+import Modal from "../components/Modal"
+import query from "../helpers.js";
+
+/**
+ * Modal displayed for permission management
+ */
+export default class ModalPermissions extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            inputSharer: ""
+        };
+    }
+
+    /**
+     * Ask a confirmation, delete the note and redirect to main page
+     */
+    _deleteNote() {
+        if (confirm("Delete this note?")) {
+            query(`/api/note/${this.props.note.id}/`, "DELETE").then((result) => {
+                this.props.redirectToMainPage();
+            });
+        }
+    }
+
+    _setReadOnly(readOnly) {
+        query(`/api/note/${this.props.note.id}/`, "PUT", {
+            readOnly: readOnly
+        });
+
+        // I'm not sure changing the props and notifying the parent of the change is the best thing to do
+        this.props.note.readOnly = readOnly;
+        this.props.onNoteUpdate(this.props.note);
+    }
+
+    _setPublic(isPublic) {
+        query(`/api/note/${this.props.note.id}/`, "PUT", {
+            public: isPublic
+        });
+
+        this.props.note.public = isPublic;
+        this.props.onNoteUpdate(this.props.note);
+    }
+
+
+    _addSharedUser() {
+        // Concat current sharers with the text input
+        let newSharedArray = [...this.props.note.sharedWith, this.state.inputSharer];
+        this.setState({ inputSharer: "" });
+
+        query(`/api/note/${this.props.note.id}/`, "PUT", {
+            sharedWith: newSharedArray
+        }).then((result) => {
+            // Remove bad sharers and update
+            if (result.badSharerNames.length > 0) {
+                this.setState({ inputSharer: "User not found" });
+                document.getElementById("input-sharer").select();
+                newSharedArray.splice(newSharedArray.indexOf(result.badSharerNames), 1);
+            }
+
+            this.props.note.sharedWith = newSharedArray;
+            this.props.onNoteUpdate(this.props.note);
+        });
+    }
+
+    _removeSharedUser(index) {
+        let sharedWith = this.props.note.sharedWith;
+        sharedWith.splice(index, 1);
+
+        query(`/api/note/${this.props.note.id}/`, "PUT", {
+            sharedWith: sharedWith
+        });
```

Comment:
Is this assuming all the other users in the list are correct ?

Reply:
Yes, it should not be possible to have invalid users in the list when you click on a user to delete him